src/core/mac/mac.cpp

```diff
@@ -1571,8 +1571,53 @@ void Mac::ReceiveDoneTask(Frame *aFrame, otError aError)
     SuccessOrExit(error = aFrame->ValidatePsdu());
 
     aFrame->GetSrcAddr(srcaddr);
+    aFrame->GetDstAddr(dstaddr);
     neighbor = GetNetif().GetMle().GetNeighbor(srcaddr);
 
+    // Destination Address Filtering
+    switch (dstaddr.mLength)
+    {
+    case 0:
+        break;
+
+    case sizeof(ShortAddress):
+        aFrame->GetDstPanId(panid);
+        VerifyOrExit((panid == kShortAddrBroadcast || panid == mPanId) &&
+                     ((mRxOnWhenIdle && dstaddr.mShortAddress == kShortAddrBroadcast) ||
+                      dstaddr.mShortAddress == mShortAddress), error = OT_ERROR_DESTINATION_ADDRESS_FILTERED);
+
+#if OPENTHREAD_FTD
+
+        // Allow non link-local multicasts from neighbor routers if FFD
+        if (neighbor == NULL && dstaddr.mShortAddress == kShortAddrBroadcast &&
+            srcaddr.mLength == sizeof(ShortAddress) &&
```

Comment:
I understand this predicate is meant to only accept non-link-local multicasts.  Strictly speaking, however, I don't think the Thread spec prohibits the use of short MAC source address with link-local multicast messages.  In any case, is this check really necessary to include?

Reply:
That was intended to be a confirmation that the below ```IsActiveRouter``` and ```GetRouterId``` functions would work, so it was more important that it didn't falsely filter out frames that should be received. I do agree that it perhaps over constrains this part, as the only frames that MUST be discarded are multicasts, so I'll tidy up